core/src/sumeragi/network_topology.rs

```diff
@@ -234,4 +356,202 @@ mod tests {
         topology.update_peer_list(new_peers);
         assert_eq!(extract_ports(&topology), vec![0, 2, 5, 7])
     }
+
+    #[test]
+    fn filter_by_role() {
+        let key_pairs =
+            core::iter::repeat_with(|| KeyPair::generate().expect("Failed to generate key pair"))
+                .take(7)
+                .collect::<Vec<_>>();
+        let mut key_pairs_iter = key_pairs.iter();
+        let peers = peers![0, 1, 2, 3, 4, 5, 6: key_pairs_iter];
+        let topology = Topology::new(peers.clone());
+
+        let dummy = "value to sign";
+        let signatures = key_pairs
+            .iter()
+            .map(|key_pair| SignatureOf::new(key_pair.clone(), &dummy).expect("Failed to sign"))
+            .collect::<Result<SignaturesOf<_>, _>>()
+            .expect("Failed to create `SignaturesOf`");
+
+        let leader_signatures =
+            topology.filter_signatures_by_roles(&[Role::Leader], signatures.iter());
+        assert_eq!(leader_signatures.len(), 1);
+        assert_eq!(leader_signatures[0].public_key(), peers[0].public_key());
+
+        let proxy_tail_signatures =
+            topology.filter_signatures_by_roles(&[Role::ProxyTail], signatures.iter());
+        assert_eq!(proxy_tail_signatures.len(), 1);
+        assert_eq!(proxy_tail_signatures[0].public_key(), peers[4].public_key());
+
+        let validating_peers_signatures =
+            topology.filter_signatures_by_roles(&[Role::ValidatingPeer], signatures.iter());
+        assert_eq!(validating_peers_signatures.len(), 3);
+        assert!(validating_peers_signatures
+            .iter()
+            .all(|s| peers[1..4].iter().any(|p| p.public_key() == s.public_key())));
```

Comment:
This isn't exactly checking for deterministic topologies. It might also miss some cases where the peers all have the same public key (which is a problem).